### DISCUSSION TITLE
Display total sum of relationship in resource table

### DISCUSSION THREAD

**Discussion by madsem** [filamentphp/filament#2161] [Help]
How could I go about displaying the SUM of a relationship in a resource table, filtered by start & end date?

I'm trying to build BI reports for a business (Total spend, total order value etc) using Filament.

For example: The Filament demo has the Orders Resource... Now I would like to create a table that would display total order value by date.

Is that doable?

**Comment by ryangjchandler** [comment]
You can resolve the value of a column yourself using `getStateUsing`

**Reply by madsem** [reply]
Thanks @ryangjchandler, but that wouldn't group the results in the resource table, right?

What I need is to show the SUM of a relationship, by date.

So real BI reports, that group & sum/avg rows.

Is that something Filament can do?

**Reply by madsem** [reply]
so the table should look something like:

date |  total orders | total order value | total revenue | total advertising spend

**Reply by ryangjchandler** [reply]
@madsem Ah, it sounds like you want a custom table with a custom query then.

**Reply by danharrin** [reply]
Well you'd need a `withSum(['relationship_name' => fn ($query) => ...])` clause on the `getTableQuery()` method of the list page, that will allow you to then use a column like `relationship_name_sum`, right?

**Reply by simonbuehler** [reply]
@madsem did you find a solution? i'm about to have the same kind of requirement.

**Comment by madsem** [comment]
Sadly no, I switched to use the table & form builder without the admin.

But in any case, there is a method called (I believe) `getContentFooter()` that has access to the model collection.
Problem there is then ofc that you need to loop through all entries.

It's def. something `imho` that is missing from filament.

But maybe it's possible to pass other arguments to that method, like an array with pre-computed totals? I don't know

**Reply by studiowizjo** [reply]
How would you like it to display? As a statistic above the table, as a column or any other way?

![1](https://user-images.githubusercontent.com/3818874/172239560-9b77efee-add5-4af0-a07b-722ac6139266.png)

If you need something similiar to this, I can share my solution. It shows statistic base on selected filters.

**Reply by eorenstein1** [reply]
@studiowizjo

I'd love to see your solution.

**Reply by studiowizjo** [reply]
1. [Create resource widget](https://filamentphp.com/docs/2.x/admin/resources/widgets#creating-a-resource-widget)

2. After each change of data in table emit event:
```php
protected function getTableQuery(): Builder
{
  $this->emit('updateSettlementsMonthSummaryWidget', [
    'month' => $this->tableFilters['date_range']['month'] ?? date('m'),
    'year' => $this->tableFilters['date_range']['year'] ?? date('Y'),
  ]);

  return Settlement::query();
}
```

3. Listen for event in your widget:

```php
protected $listeners = ['updateSettlementsMonthSummaryWidget' => 'update'];

public array $monthsSum;

public function update($dateRange)
{
  //some logic to calculate $this->monthsSum
}

protected function getCards(): array
{
  return [
    Card::make('Sum for selected date range', number_format($this->monthsSum[0], 2, ',', '.') . ' PLN')
      ->description($this->getDescription())
      ->descriptionIcon($this->getDescriptionIcon())
      ->chart($this->monthsSum)
      ->color($this->getColor()),
  ];
}
```

However, [there is one disadvantage](https://github.com/laravel-filament/filament/discussions/2552) of this approach - data in widget refreshes, but chart doesnt.

**Reply by danharrin** [reply]
We should look at ways of using stats / charts outside of widgets, so that you can have the data available from the table easily (since its part of the same LW component).

**Reply by studiowizjo** [reply]
Maybe do something like TableWidgets?
